### ISSUE TITLE
"save file" and "confirm close" dialogs

### ISSUE THREAD

**Issue by kzidane** [cs50/harvard.cs50.droplet#78]
Should we perhaps merge the "confirm close" and "save file" dialogs into one if possible? If not, should we have "confirm close" show up before "save file"? Much like alerts, modal dialogs might be a little annoying, and I assume choosing not to save changes implicitly means I'm ok with losing dangling pieces?

![save-file](https://user-images.githubusercontent.com/7230211/29996106-6b27e958-8fc6-11e7-941a-cbab9b0b5a64.png)

![confirm-close](https://user-images.githubusercontent.com/7230211/29996107-6cd233e4-8fc6-11e7-93b6-f46b55d020bb.png)

CC @dmalan

**Comment by crossroads1112** [comment]
@kzidane I'm not sure the "these pieces will disappear" message is even warranted any more since, as we saw today, they don't actually disappear. I'd say we should just get rid of it entirely.

**Comment by dmalan** [comment]
On the other hand, should we just delete the orphaned blocks as warned?

**Comment by kzidane** [comment]
@crossroads1112 I think #78 refers to closing the tab rather than just switching to code and back? In case you close the tab, the dangling blocks disappear, yes?

@dmalan I personally think keeping them would be nice if there are no down sides. Perhaps we just need to change the warning dialog that appears when switching to text to say that any dangling blocks won't show up in text mode instead of saying that they would disappear?

**Comment by dmalan** [comment]
What would be the implications of just showing the orphaned code as a /* */ block of source at the bottom of the file?

**Comment by kzidane** [comment]
@dmalan no, I just meant keeping them just in the droplet stage not in the code itself.

**Comment by dmalan** [comment]
Right, but what would be the implications of just showing the orphaned code as a /* */ block of source at the bottom of the file? :p

**Comment by kzidane** [comment]
Presumably students would be able to peek at and possibly manually uncomment and cut and paste somewhere which would be easier by dragging and dropping in Droplet anyway if that's what you're asking?

And of course we'd need a way to distinguish between actual `/**/` comments at the bottom of the file and comments representing dangling blocks unless we assume any `/**/` comment at the bottom of a file is a dangling block even if added manually?

Alternatively we can perhaps consider *any* `/**/` comment to be a dangling block such that when a student drags a block out we could just `/**/`-comment this block where it is, but we'd perhaps need a way to distinguish between `/**/`-commented code and normal `/**/` comments if that might work.

**Comment by crossroads1112** [comment]
I don't think we should turn floating blocks into comments for a couple of reasons:

* It would require us to extend the C language by imbuing certain comments with semantic meaning. Not that there isn't precedent for this (doxygen for example), but I don't love the idea of doing this the first time students see C. 
* I'm not sure that it really adds that much. True, it lets us preserve floating blocks even after the document is closed and extends the parallels between the block view and the text view, but I don't think the former is that impactful and it wouldn't even entirely solve the latter since the positions of the floating blocks would still be lost. Even if the floating block were placed next to some block of code at the top of the file, the comment would appear all the way at the bottom, and there'd be no way (without encoding more information in the comment) to restore that position when we go back into blocks mode.

If we wanted to preserve floating blocks even when closing the document, we could do it the same way that ace stores undo history by storing JSON-encoded metadata for each file in `~/.c9`. Not sure this is really worth doing though.
